### PR TITLE
Tensor parallel  distributed strategy without using deepspeed

### PR DESCRIPTION
**Pull request by kalyanjk**
Tensor parallel by extending GaudiLlamaAttention -> TPGaudiLlamaAttention  and GaudiLlamaMLP -> TPGaudiLlamaMLP

use parameter --distributed_strategy="tp" to invoke this code path

### CODE REVIEW

File: optimum/habana/distributed/strategy.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+from abc import abstractmethod
+from typing import Any, List, Mapping
+
+import torch
+import torch.distributed
+from torch import nn
+
+#from optimum.habana.distributed import tp_wrapping
```
Comment (msinnha1) [inline optimum/habana/distributed/strategy.py:8]: minor: please  remove the commented code

File: optimum/habana/distributed/strategy.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    @abstractmethod
+    def distribute_layer(self, block: nn.Module, layer: int) -> nn.Module:
+        """
+        Distribute each layer as-appropriate
+        """
+        pass
+
+
+class NotDistributed(DistributedStrategy):
```
Comment (msinnha1) [inline optimum/habana/distributed/strategy.py:32]: why derived class is NotDistributed for the base class as DistributedStrategy? It is creating some confusion in readability, may require some other name?

File: optimum/habana/distributed/strategy.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                layer_id = layer_id + 1
+            if remainder > 0:
+                self.layer_to_device[layer_id] = devices[dev_idx]
+                layer_id = layer_id + 1
+                remainder -= 1
+
+    def distribute_layer(self, block: nn.Module, layer: int) -> nn.Module:
+        device = self.layer_to_device[layer]
+        if self.from_meta:
+            # https://github.com/pytorch/pytorch/pull/113647
```
Comment (msinnha1) [inline optimum/habana/distributed/strategy.py:94]: this PR is closed, and we can possibly remove the reference to such comments from foundation repo, #Comment

File: optimum/habana/distributed/tensorparallel.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+def apply_colwise_tp(par_mod: nn.Linear, mod: nn.Linear, world_size, rank):
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = mod.out_features // world_size
+    with torch.no_grad():
+        par_mod.weight.copy_(
+            torch.split(mod.weight, output_size_per_partition, dim=0)[rank]
+        )
+        if par_mod.bias is not None:
+            par_mod.bias.copy_(torch.split(mod.bias, output_size_per_partition)[rank])
+    # print(f"For rank {rank}, we have the following weights: Base weight {mod.weight} bias {mod.bias}; Par weight {par_mod.weight}, bias {par_mod.bias}")
```
Comment (msinnha1) [inline optimum/habana/distributed/tensorparallel.py:20]: #Comment

File: optimum/habana/distributed/tensorparallel.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    with torch.no_grad():
+        par_mod.weight.copy_(
+            torch.split(mod.weight, output_size_per_partition, dim=1)[rank]
+        )
+        if par_mod.bias is not None:
+            if rank == 0:
+                par_mod.bias.copy_(mod.bias)
+            else:
+                par_mod.bias.zero_()
+    # print(f"For rank {rank}, we have the following weights: Base weight {mod.weight}, bias {mod.bias}; Par weight {par_mod.weight}, bias {par_mod.bias}")
```
Comment (msinnha1) [inline optimum/habana/distributed/tensorparallel.py:35]: #Comment

File: optimum/habana/distributed/tensorparallel.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def apply_embedding_tp(par_mod: nn.Embedding, mod: nn.Embedding, world_size, rank):
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = mod.embedding_dim // world_size
+    with torch.no_grad():
+        par_mod.weight.copy_(
+            torch.split(mod.weight, output_size_per_partition, dim=1)[rank]
+        )
+    # print(f"For rank {rank}, we have the following weights: Base weight {mod.weight} bias {mod.bias}; Par weight {par_mod.weight}, bias {par_mod.bias}")
```
Comment (msinnha1) [inline optimum/habana/distributed/tensorparallel.py:45]: #Comment

File: optimum/habana/distributed/tensorparallel.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    """Gather the input tensor across model parallel group."""
+    world_size = dist.get_world_size()
+
+    if world_size == 1:
+        return input_
+
+    # The transposes here are to avoid excessive recompilation due to split()
+    # specializing the dimension where the all_gather is happening
+    last_dim = input_.dim() - 1
+    # Starting PT 2.3, we can go back to funcol.all_gather_tensor
```
Comment (msinnha1) [inline optimum/habana/distributed/tensorparallel.py:119]: #Comment

**Review by msinnha1 — COMMENTED** [review]
This is a big patch and reviewing it further

**Review by msinnha1 — APPROVED** [review]

**Review by dvarshney-habana — APPROVED** [review]

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 8]
```diff
@@ -1013,6 +1161,7 @@ def forward(
             global has_fused_rope
             has_fused_rope = False
 
+
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: > minor: please remove this

Done

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 9]
```diff
@@ -731,17 +870,23 @@ def __init__(self, config: LlamaConfig):
...
         self.vocab_size = config.vocab_size
-
+        self.distributed_strategy = config.distributed_strategy
+        config.distributed_strategy = None
         self.embed_tokens = torch.nn.Embedding(config.vocab_size, config.hidden_size, self.padding_idx)
-        self.layers = torch.nn.ModuleList(
-            [GaudiLlamaDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
-        )
+        layers = []
+        for i in range(config.num_hidden_layers):
```
[the remark this answers is not in the document]
Comment (kalyanjk) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: > minor: layer_idx in place of 'i'

Done

File: optimum/habana/distributed/strategy.py (hunk 1 above)
Comment (kalyanjk) [reply optimum/habana/distributed/strategy.py]: > minor: please remove the commented code

Done

File: optimum/habana/distributed/strategy.py (hunk 3 above)
Comment (kalyanjk) [reply optimum/habana/distributed/strategy.py]: Done

**Review by dvarshney-habana — APPROVED** [review]

**Comment by astachowiczhabana** [comment]
https://github.com/huggingface/optimum-habana/pull/1121

**Pull request merged**